## Mental Data KNN ML

[Colab NoteBook](https://colab.research.google.com/drive/1Xwi9BtigmWb3NScq2QeDlX0ARBTlo1fC?usp=sharing)

[Github for full](https://github.com/z5208980/machine-learning-health/tree/main/mental_health) (dataset, model, notebook)

In [107]:
import numpy as np
import pandas as pd
import pickle

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn import metrics
from sklearn.metrics import accuracy_score, mean_squared_error, precision_recall_curve
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsClassifier

train_df = pd.read_csv('https://raw.githubusercontent.com/z5208980/machine-learning-health/main/mental_health/data/processed.csv')
# train_df.info()

In [108]:
df = pd.read_csv('https://raw.githubusercontent.com/z5208980/machine-learning-health/main/mental_health/data/processed.csv')

# improved from ~89 to ~95
scaler = MinMaxScaler()
df['Age'] = scaler.fit_transform(df[['Age']])

In [109]:
feature_cols = ['Age', 'Gender', 'family_history', 'treatment', 'work_interfere', 'tech_company', 'seek_help', 'leave', 'mental_health_interview', 'phys_health_interview']
X = df[feature_cols]
y = df.treatment

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=200)

In [110]:
def Knn():
    model = KNeighborsClassifier(n_neighbors=3, weights='distance')
    model.fit(X_train, y_train)
    
    y_pred_class = model.predict(X_test)

    print('RESULT')
    print('Accuracy:', metrics.accuracy_score(y_test, y_pred_class))
    print('Precision:', metrics.precision_score(y_test, y_pred_class))
    print('AUC Score:', metrics.roc_auc_score(y_test, y_pred_class))
    print('Cross-validated AUC:', cross_val_score(model, X, y, cv=10, scoring='roc_auc').mean())

    filename = '/content/sample_data/model.sav'
    pickle.dump(model, open(filename, 'wb'))

Knn()

RESULT
Accuracy: 0.9714285714285714
Precision: 0.9743589743589743
AUC Score: 0.9714182052729178
Cross-validated AUC: 0.990626165153935



Using Knn (KNeighborsClassifier)
- Score:  0.9714285714285714
- Params:  `{'weights': 'distance', 'n_neighbors': 3}`




## DEMO

KNN model can be found here [Github](https://github.com/z5208980/machine-learning-health/blob/main/mental_health/model.sav
)

Input values requires
- Age
- Gender
- family_history
- treatment
- work_interfere
- tech_company
- seek_help
- leave
- mental_health_interview
- phys_health_interview

In [111]:
model = pickle.load(open('/content/sample_data/model.sav', 'rb'))   # load model

# ['Age', 'Gender', 'family_history', 'treatment', 'work_interfere', 'tech_company', 'seek_help', 'leave', 'mental_health_interview', 'phys_health_interview']
val = []
row = 70
for x in X_train.iloc[row]:
  val.append(x)

input = [val]
output = model.predict(input)

print("X=%s, Predicted=%s, Actually=%s" % (input[0], output[0], y_train.iloc[row]))

X=[0.36363636363636365, 1.0, 0.0, 0.0, 1.0, 1.0, 1.0, 4.0, 1.0, 0.0], Predicted=0, Actually=0


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  "X does not have valid feature names, but"
